design/how-does-data-distribution-work.md

```diff
@@ -0,0 +1,83 @@
+
+## Components
+
+**Storage server (struct TCServerInfo):** DD creates a TCServerInfo object for each storage server (SS). The TCServerInfo includes: (i) the SS’ locality, which includes the processID that is unique to ip:port, the zoneId that specifies which rack the SS is on, and the dcId that specifies which DC the SS is in; (ii) the server’s teams, which will be discussed in the following paragraph; (iii) the tracker that monitor the status of the server; and (iv) extra information related to the server’s interface and preference. A server is healthy if its storage engine on the process is the same with the configured storage engine, and it is marked as desired by DD.
+
+**Machine (struct TCMachineInfo)**: A machine in FDB is considered as a rack, because a typical FDB cluster will only use one physical host from each rack in the datacenter to reduce the impact of regular rack-maintenance events on the cluster. All servers on the same rack belong to the same machine. A machine is healthy if there exists a healthy server on the machine.
+
+**Server team (struct TCTeamInfo)**: A server team is a group of k servers that host the same key ranges, where k is the replication factor that is usually three. A server team is healthy if every server in the team is healthy and those servers’ localities satisfies the replication requirement. Servers are grouped into server teams to reduce the possibility of data unavailability events at the event of k server failures. 
+
+**Machine team (struct TCMachineTeamInfo)**: A machine team is a group of k machines, where k is the replication factor. Each server team must be on a machine team, meaning that each server in the server team is on a machine in the machine team and that no two servers are on the same machine. Similar to the purpose of server teams, machine teams are used to reduce the possibility of data unavailability events at the event of k _machine_ failures. A machine team is healthy if every machine on the team is healthy and machines’ localities satisfy the replication policy.
+
+**TeamCollection**: It has a global view of all servers and server teams, machines and machine teams. With the information, it creates server teams and machine teams. It also maintains the configuration setting for DD, which is used to create teams and decide which type of storage servers to recruit.
+
+**Shard (struct DDShardInfo)**: A shard is a key range. A shard is maintained by a server team. A server team is responsible for lots of shards. Each shard has similar amount of data. When a shard has too much data or has too much write traffic, it will be split into multiple shards and redistributed to server teams. Likewise, when a shard has too small data, it can be merged with its neighbors.
```

Comment:
```suggestion
**Shard (struct DDShardInfo)**: A shard is a key range. A shard is maintained by a server team. A server team is responsible for many shards. Each shard has a similar amount of data. When a shard has too much data or has too much write traffic, it will be split into multiple shards and redistributed to server teams. Likewise, when a shard has too little data, it can be merged with its neighbors.
```